File: apps/mobile/android/app/src/main/java/io/chatic/dou/module/PushMarksModule.kt
package io.chatic.dou.module

import com.facebook.react.bridge.Arguments
import com.facebook.react.bridge.Promise
import com.facebook.react.bridge.ReactApplicationContext
import com.facebook.react.bridge.ReactContextBaseJavaModule
import com.facebook.react.bridge.ReactMethod
import io.chatic.dou.push.PushMarkStore

/**
 * Drains the cross-cloud push marks a background chat push recorded (ADR-0056) — read and clear
 * in one native call, so a mark reaches the web exactly once. Called on boot (after `WebAppReady`)
 * and on foreground return; see `useFcmHandler`'s `FetchPushMarks` handler on the JS side.
 */
class PushMarksModule(reactContext: ReactApplicationContext) : ReactContextBaseJavaModule(reactContext) {

    override fun getName(): String {
        return "PushMarks"
    }

    @ReactMethod
    fun drain(promise: Promise) {
        try {
            val marks = PushMarkStore.drainAll(reactApplicationContext)
            val result = Arguments.createArray()
            marks.forEach { record ->
                val map = Arguments.createMap()
                record.forEach { (key, value) -> map.putString(key, value) }
                result.pushMap(map)
            }
            promise.resolve(result)
        } catch (e: Exception) {
            promise.reject("PUSH_MARKS_DRAIN_FAILED", e.message)
        }
    }
}

File: apps/mobile/ios/Bridges/PushMarksModule.m
#import <React/RCTBridgeModule.h>

/**
 * Drains the cross-cloud push marks the Notification Service Extension recorded for a background
 * chat push (ADR-0056) — read and clear the shared App Group store in one call, so a mark reaches
 * the web exactly once. Called on boot (after `WebAppReady`) and on foreground return; see
 * `useFcmHandler`'s `FetchPushMarks` handler on the JS side.
 *
 * App Group id and key must match `NotificationService.swift`'s `appGroupId`/`pushMarksKey`.
 */
@interface PushMarksModule : NSObject <RCTBridgeModule>
@end

@implementation PushMarksModule

RCT_EXPORT_MODULE(PushMarks);

- (dispatch_queue_t)methodQueue {
    return dispatch_get_main_queue();
}

RCT_EXPORT_METHOD(drain:(RCTPromiseResolveBlock)resolve
                  reject:(RCTPromiseRejectBlock)reject) {
    NSString *appGroupId = @"group.io.chatic.dou";
    NSString *marksKey = @"push_marks";

    NSUserDefaults *defaults = [[NSUserDefaults alloc] initWithSuiteName:appGroupId];
    if (!defaults) {
        resolve(@[]);
        return;
    }

    NSArray *marks = [defaults arrayForKey:marksKey] ?: @[];
    [defaults removeObjectForKey:marksKey];
    resolve(marks);
}

@end
